contrib/lkt/extensions/src/liblktlang-implementation-extensions.adb

```diff
@@ -162,6 +162,14 @@ package body Liblktlang.Implementation.Extensions is
          Charset => "ascii",
          Reparse => False);
    begin
+      if Node.Kind = lkt_Import and I.AST_Root = null then
```

Comment:
I'm not sure there's any benefit in restraining the diagnostic emission to the case where `Node` is of kind `lkt_Import`. Do you have anything in mind? My intuition is that if we ever need to call this property from another kind of node, we will probably want the same kind of error handling (or it will be easy to refine the condition then).

Reply:
um... I just wanted to ensure that this specific diagnostic is only added for import nodes but if I understand what you're saying, it is possible to have this error on another node than a lkt_import kind one?

Reply:
ping @Roldak

Reply:
I’ve just realized this violates the design of diagnostics, which are for lexing/parsing errors (as described in the error message from `lkt_resolve.py`), while here we have an error happening during semantic analysis.

@raph-amiard recently introduce a mechanism precisely to be notified when a file is missing (see event handlers, from U712-008). Unfortunately this mechanism is not yet available from the Python API.

Hm… what about re-implementing `lkt_resolve.py` in Ada, to use the proper abstraction for this?

Reply:
> I’ve just realized this violates the design of diagnostics, which are for lexing/parsing errors

I'm not sure this was ever a real design point, it's just that so far they've only been used for that. Now, you might have reasons to think it's better to keep them for that ?

> @raph-amiard recently introduce a mechanism precisely to be notified when a file is missing (see event handlers, from U712-008). Unfortunately this mechanism is not yet available from the Python API.

This mechanism unfortunately doesn't yet include the "origin" of the signal, ie. the import here, so wouldn't be enough even in Ada.

> Hm… what about re-implementing lkt_resolve.py in Ada, to use the proper abstraction for this?

We already have that, it's lkt_toolbox... TBH I thought all lkt tests were already using it ... Just realized it's using lkt_resolve.py. 

But it's not a great solution anyway: In Langkit we'll use lkt from python, so we need a solution that works in Python.

Reply:
> > I’ve just realized this violates the design of diagnostics, which are for lexing/parsing errors
> 
> I'm not sure this was ever a real design point, it's just that so far they've only been used for that. Now, you might have reasons to think it's better to keep them for that ?

Yes: if you load unit A that tried to import unit B, the error message will go to unit A’s diagnostics. However if the loading of B is succesful, but this unit fails to load unit C, the diagnostic will go to unit B, so you won’t see looking at unit A that there is an error. So it looks like we need something global.

> This mechanism unfortunately doesn't yet include the "origin" of the signal, ie. the import here, so wouldn't be enough even in Ada.

I see, but that’s probably a sign that we should extend this API to include the origin.

> We already have that, it's lkt_toolbox... TBH I thought all lkt tests were already using it ... Just realized it's using lkt_resolve.py.
> 
> But it's not a great solution anyway: In Langkit we'll use lkt from python, so we need a solution that works in Python.

Ok… IMO this means we should tackle this and bind this API in Python. I’ll have a look at the feasability.

Reply:
Thinking about this further, I think this diagnostic should simply be emited by the `p_check_semantic`/`p_check_legality` properties of the DSL, as part of the normal error reporting process, and that it's pretty straightforward to do that ...

One thing that Romain raised is that on this error message we probably want to stop further analysis, or at  least display, since you're going to get a lot of garbage error messages, and Laurent's current solution has this benefit.

But this can probably be done if we really want to, I'll check to see how easy it would be

Reply:
Ah yes, that’s even better! I forgot about the existing semantic checks machinery in Lkt… I’m much more in favor of that!

Reply:
Rework done!